checker/src/org/checkerframework/checker/lock/LockStore.java

```diff
@@ -137,4 +159,83 @@ protected void internalDotOutput(StringBuilder result) {
                 + "\\n");
         super.internalDotOutput(result);
     }
+
+    @Override
+    protected boolean isSideEffectFree(AnnotatedTypeFactory atypeFactory,
+            ExecutableElement method) {
+        return ((LockAnnotatedTypeFactory) atypeFactory).methodSideEffectAnnotation(method, false) == SideEffectAnnotation.RELEASESNOLOCKS ||
+               super.isSideEffectFree(atypeFactory, method);
+    }
+
+    // Same as super method, except for the call to localVariableValues.clear();
+    // See the comment next to the call for more details.
+    @Override
+    public void updateForMethodCall(MethodInvocationNode n,
+            AnnotatedTypeFactory atypeFactory, CFValue val) {
+        ExecutableElement method = n.getTarget().getMethod();
+
+        // case 1: remove information if necessary
+        if (!(((LockChecker) analysis.getTypeFactory().getContext()).hasOption("assumeSideEffectFree")
```

Comment:
But isn't the check for "assumeSideEffectFree" part of your changes?
You are also overriding "isSideEffectFree" and I'm wondering why you're not checking for the option there to keep the changes even smaller.